### ISSUE TITLE
GoNode shows incorrect keywords

### ISSUE THREAD

**Issue by chpicone** [geosolutions-it/oaw-geonode#37]
OAW reported an issue with the keywords:
- Advanced search page: 
the “keywords” section shows no available keywords:
![image.png](https://images.zenhubusercontent.com/5e98219c18069195a59f5217/7c0f4ba9-827b-416a-8e53-93b6e44766a6)

- Layers page: 
the “keywords” section shows 2 available keywords (but we add 2 different keywords to the AC04078710 layer and they were not in the list):
![image.png](https://images.zenhubusercontent.com/5e98219c18069195a59f5217/18996ceb-2b54-4159-9e6e-439d91385d45)

**Comment by mattiagiupponi** [comment]
### Analysis:

After some analysis, the issue is caused by this function [dump_bulk_tree](https://github.com/GeoNode/geonode/blob/971e71613e7c3cbb59cc6393d060b4df2bb260cf/geonode/base/models.py#L344
)

In few words, if we assign a keyword to a resource that is a child of another `HierarchicalKeyword` without assigning the parent keyword to the resource, the function [silently ignores](https://github.com/GeoNode/geonode/blob/971e71613e7c3cbb59cc6393d060b4df2bb260cf/geonode/base/models.py#L407
) the exception when it tries to reproduce the [keyword tree](https://github.com/GeoNode/geonode/blob/971e71613e7c3cbb59cc6393d060b4df2bb260cf/geonode/base/models.py#L402).


Example:
In the `HierarchicalKeyword` admin page we have this situation:

![image](https://user-images.githubusercontent.com/51856725/123836214-b9cbb400-d909-11eb-9039-dfd5ba91e411.png)

We decided to assign to a document the keyword `infoMapAccessService` that is a child of `Monitoring stations (Water Framework Directive)`, so we will have this situation in the document keywords:

![image](https://user-images.githubusercontent.com/51856725/123836768-59894200-d90a-11eb-935f-142d27efd44f.png)

When we refresh the page (or call the keywords [API](http://localhost:8000/h_keywords_api?type=document)) it returns an empty list of keywords:

![image](https://user-images.githubusercontent.com/51856725/123836886-80477880-d90a-11eb-9ca2-84a620125a2e.png)

And from the console log (after adding a print) it will raise a `KeyError` because he tries to find into the `lnk ` dict the parent. Plus it stops completely populating the JSON with the keywords.

As counter-proof, if we add also the parent keyword to the document like this:

![image](https://user-images.githubusercontent.com/51856725/123837310-fea41a80-d90a-11eb-8c54-a7cd17732654.png)

The endpoint will show correctly the keywords tree.

![image](https://user-images.githubusercontent.com/51856725/123837492-314e1300-d90b-11eb-8d4d-7a0129599e9f.png)

### Solution:

We should completely revise the algorithm that generates the keywords payload to avoid those kinds of errors and generate the payload even if the parent `keyword` is not assigned to the same resource as the child one. 
Plus we should also decide if:
- Return the payload without the tree (as the` api/v2` is already doing)
- Return the payload with the tree by adding also the parent keywords even if the tag_count is 0.
Nice to have:
The [keyword facet](https://github.com/GeoNode/geonode/blob/971e71613e7c3cbb59cc6393d060b4df2bb260cf/geonode/urls.py#L125) is not passing through the `Tastypie` endpoints or from the `api/v2` endpoint, should be better to move this endpoint to the API ones

@giohappy @chpicone any thought?

**Comment by giohappy** [comment]
@chpicone if it's not super urgent I would make Syed work on this. He will start from zero, so it will take a while to rump up. Let me know.

**Comment by giohappy** [comment]
I add just one note to the analysis from @mattiagiupponi, then I will add my proposal to fix this.

The problem is two-fold. From one side we have what @mattiagiupponi has described, in case the parent of a keyword is not assigned to a resource the dump tree method fails silently. Moreover the algorithm assumes that the children of a keyword have been created after their parent, because keywords are looped in insertion order. If the parent of a keyword comes after it it won't be available [inside the lnk list](https://github.com/GeoNode/geonode/blob/971e71613e7c3cbb59cc6393d060b4df2bb260cf/geonode/base/models.py#L402), and the algorithm will fail.

Since facets as we know them will disappear on master I wouldn't bother to much with a refactoring. I would simply fix the [dump_bulk_tree](https://github.com/GeoNode/geonode/blob/8462948d00585ed7c3b25081670ae9af0c43048b/geonode/base/models.py#L344) method to return the whole keyword tree even though parents of a resource keyword are not explicetly assigned to the resource itself.
I think this is semanthically correct, since the keywords hierarchy is a concept internal to the keywords data structure, independently from resources.

**Comment by asgharsir** [comment]
I haven't looked into the code yet, rather I have started using the app as a user, and that's just now, so excuse me for my little knowledge. But I can't reproduce the above. I can see the keywords on both pages.
![image](https://user-images.githubusercontent.com/47123144/124634501-53e9aa00-de7e-11eb-8c23-ffb7fb6586d4.png)

**Comment by asgharsir** [comment]
Update. I could produce such error. Observation is even different from @mattiagiupponi 's one. In my tests, no keywords appear when you add a child, whether or not, parent keyword is added to the resource. And whether or not, parent or child is associated with a resource. The difference in the observation might have something to do with the order of insertion as pointed out by @giohappy .

**Comment by chpicone** [comment]
@asgharsir yes correct but also the hierarchical levels are to be considered.
I mean if we have all keywords at the root level the order of insertion is not important. The order is important when we have keywords under others. In that case the algorithm need to scan before keywords at top levels and then to the lower levels, otherway they are not considered.

**Comment by asgharsir** [comment]
After inspection of the code I can verify that issue needs fix in dump_bulk_tree method as pointed out by @mattiagiupponi earlier. Further that bug would impact irrespective of parent's tagging, if a child is tagged.

I have apparently fixed it to the extent of manual testing, but I am blocked for adding tests before some confirmation. I can see that an existing test function geonode.layers.tests.LayersTest.test_layer_save and the whole geonode.layers.tests.LayersTest class, which tests this method, is failing locally. If someone may confirm the status of the earlier tests.

Secondly, as discussed earlier that this issue impacts the core geonode, I am asked to request @chpicone to create an issue under geonode core for this. After that I may have a branch name accordingly.
![image](https://user-images.githubusercontent.com/47123144/124758032-193b4c80-df26-11eb-8324-812ea03e4f68.png)

**Comment by mattiagiupponi** [comment]
@asgharsir I can confirm that even on CircleCI that specific tests if failing `geonode.layers.tests.LayersTest.test_layer_save` so your fix does not create any kind of regression on that specific test. The other tests from that class, are working.
If you can fix it in your PR will be perfect.
Please create also some specific unit tests for the function in order to be confident with the fix.

**Comment by asgharsir** [comment]
Thanks. Yes, only this confirmation was a blocker for writing new tests.  But writing tests for this method will involve understanding two new libraries. So, might take a bit of time.

**Comment by chpicone** [comment]
@asgharsir @mattiagiupponi tahnks for your support.
Do you think possible to have a PR ready by tomorrow?

**Comment by asgharsir** [comment]
I am confident that, yes. I am also looking to fix the issue of the earlier tests with this method. 

I thought I could have made the PR even today. But I am still not very comfortable with the working environment the way it is installed. Struggling my way through the debug running and testing without any command line fluency. Paver thing didn't work.

**Comment by asgharsir** [comment]
The fix is ready and unit tested. Seeing the issues in the previous testing I approached for using far lighter APITestCase and avoided the GeoNodeBaseTestSupport. Created factories in the base app, which I have not only used for this test but it also worked for the earlier test case which failed for the 'exotic encoding text'. This not only worked for present and past cases, but we can use the factories to avoid, wherever possible, in future the GeoNodeBaseTestSupport which is far heavier, and 'create_layer_data()' which is 'hard-coded' data so also obscure the data for readability of the tests.

I have not deleted or modified the existing failing test 'geonode.layers.tests.LayersTest' because it is actually going in the flow of a data dependency, as mentioned above. I just tested the failing part in the new tests.

Importantly also, I was not only slowed by third party libraries impacting the code, specially the models, and also because I have yet to explore this code base and its development environment a lot, but also for undocumented facts, as important as knowing the cleaning and making public of a resource, which was a mandatory knowledge for testing this.

And now, @chpicone @mattiagiupponi , should I continue with the branch name "issue-37"? And which branch would be appropriate to make a PR in?

**Comment by mattiagiupponi** [comment]
@chpicone can you create the issue on GeoNode core repository? so he can change the branch name and follow the standard branch/commits naming convention.

**Comment by chpicone** [comment]
Issue created:
https://github.com/GeoNode/geonode/issues/7750
@giohappy can you assign @asgharsir to this issue (I cannot do it).

**Comment by giohappy** [comment]
@asgharsir added to the GeoNode Team and issue assigned.
